src/differentiation/jaches_products.jl

```diff
@@ -198,112 +198,105 @@ end
 
-function LinearAlgebra.mul!(dy::AbstractVector, L::HesVec, v::AbstractVector)
-    if L.autodiff
-        numauto_hesvec!(dy, L.f, L.x, v, L.cache1, L.cache2, L.cache3)
-    else
-        num_hesvec!(dy, L.f, L.x, v, L.cache1, L.cache2, L.cache3)
-    end
-end
+    vecprod  = autodiff ? auto_jacvec  : num_jacvec
+    vecprod! = autodiff ? auto_jacvec! : num_jacvec!
 
-struct HesVecGrad{G, T1, T2, uType}
-    g::G
-    cache1::T1
-    cache2::T2
-    x::uType
-    autodiff::Bool
+    L = FwdModeAutoDiffVecProd(f, u, cache, vecprod, vecprod!)
+
+    FunctionOperator(L, u, u;
+                     isinplace = true, outofplace = true,
+                     p = p, t = t, islinear = true,
+                    )
 end
 
-function HesVecGrad(g, x::AbstractArray, tag = DeivVecTag(); autodiff = false)
+function HesVec(f, u::AbstractArray, p = nothing, t = nothing; autodiff = true)
+
     if autodiff
-        cache1 = Dual{typeof(ForwardDiff.Tag(tag, eltype(x))), eltype(x), 1
-                      }.(x, ForwardDiff.Partials.(tuple.(x)))
-        cache2 = Dual{typeof(ForwardDiff.Tag(tag, eltype(x))), eltype(x), 1
-                      }.(x, ForwardDiff.Partials.(tuple.(x)))
+        cache1 = ForwardDiff.GradientConfig(f, u)
+        cache2 = similar(u)
+        cache3 = similar(u)
     else
-        cache1 = similar(x)
-        cache2 = similar(x)
+        cache1 = similar(u)
+        cache2 = similar(u)
+        cache3 = similar(u)
     end
-    HesVecGrad(g, cache1, cache2, x, autodiff)
-end
 
-Base.size(L::HesVecGrad) = (length(L.cache2), length(L.cache2))
-Base.size(L::HesVecGrad, i::Int) = length(L.cache2)
-function Base.:*(L::HesVecGrad, v::AbstractVector)
-    L.autodiff ? auto_hesvecgrad(L.g, L.x, v) : num_hesvecgrad(L.g, L.x, v)
+    cache = (cache1, cache2, cache3,)
+
+    vecprod  = autodiff ? numauto_hesvec  : num_hesvec
+    vecprod! = autodiff ? numauto_hesvec! : num_hesvec!
+
+    L = FwdModeAutoDiffVecProd(f, u, cache, vecprod, vecprod!)
+
+    FunctionOperator(L, u, u;
+                     isinplace = true, outofplace = true,
+                     p = p, t = t, islinear = true,
+                    )
 end
 
-function LinearAlgebra.mul!(dy::AbstractVector,
-                            L::HesVecGrad,
-                            v::AbstractVector)
-    if L.autodiff
-        auto_hesvecgrad!(dy, L.g, L.x, v, L.cache1, L.cache2)
+function HesVecGrad(f, u::AbstractArray, p = nothing, t = nothing; autodiff = true)
+
+    if autodiff
+        cache1 = Dual{
+                      typeof(ForwardDiff.Tag(DeivVecTag(), eltype(u))), eltype(u), 1
+                     }.(u, ForwardDiff.Partials.(tuple.(u)))
+
+        cache2 = copy(cache1)
     else
-        num_hesvecgrad!(dy, L.g, L.x, v, L.cache1, L.cache2)
+        cache1 = similar(u)
+        cache2 = similar(u)
     end
+
+    cache = (cache1, cache2,)
+
+    vecprod  = autodiff ? auto_hesvecgrad  : num_hesvecgrad
+    vecprod! = autodiff ? auto_hesvecgrad! : num_hesvecgrad!
```

Comment:
`f` will not necessarily support both. They have different calling conventions. The function needs to be checked before declaring this.